packages/@aws-cdk/cdk/lib/app.ts

```diff
@@ -30,42 +108,27 @@ export class App extends Root {
   }
 
   /**
-   * Runs the program. Output is written to output directory as specified in the request.
+   * @deprecated
    */
-  public run(): void {
-    const outdir = process.env[cxapi.OUTDIR_ENV];
-    if (!outdir) {
-      process.stderr.write(`ERROR: The environment variable "${cxapi.OUTDIR_ENV}" is not defined\n`);
-      process.stderr.write('AWS CDK Toolkit (>= 0.11.0) is required in order to interact with this program.\n');
-      process.exit(1);
-      return;
-    }
-
-    const result: cxapi.SynthesizeResponse = {
-      version: cxapi.PROTO_RESPONSE_VERSION,
-      stacks: this.synthesizeStacks(Object.keys(this.stacks)),
-      runtime: this.collectRuntimeInformation()
-    };
-
-    const outfile = path.join(outdir, cxapi.OUTFILE_NAME);
-    fs.writeFileSync(outfile, JSON.stringify(result, undefined, 2));
+  public run() {
+    this.node.addWarning(`It's not necessary to call app.run() anymore`);
   }
 
   /**
    * Synthesize and validate a single stack
    * @param stackName The name of the stack to synthesize
    */
   public synthesizeStack(stackName: string): cxapi.SynthesizedStack {
-    const stack = this.getStack(stackName);
-
-    this.node.prepareTree();
-
-    // first, validate this stack and stop if there are errors.
-    const errors = stack.node.validateTree();
-    if (errors.length > 0) {
-      const errorList = errors.map(e => `[${e.source.node.path}] ${e.message}`).join('\n  ');
-      throw new Error(`Stack validation failed with the following errors:\n  ${errorList}`);
+    // Maintain the existing contract that `synthesizeStack` can be called
```

Comment:
I'd still flag as `@deprecated`, but maybe that requires we devise an other mechanism for customers to get the "prepare + validate" behavior. I feel like it's wrong for user land to call `synthesizeStack` directly.